packages/muelu/test/toggletransfer/CMakeLists.txt

```diff
@@ -93,15 +93,6 @@ IF (${PACKAGE_NAME}_ENABLE_Tpetra AND ${PACKAGE_NAME}_ENABLE_Ifpack2 AND ${PACKA
     COMM serial mpi
     )
 
-  TRIBITS_ADD_TEST(
-    TogglePFactory_driver
-    NAME "Driver_TogglePFactory_semi_sa_line_easy_Tpetra"
-    ARGS "--matrixType=Laplace3D --nx=40 --ny=40 --nz=60 --mz=1 --xmlFile=parameters_semi_sa_easy.xml --linAlgebra=Tpetra"
-    NUM_MPI_PROCS 4
-    COMM serial mpi
-    )
-
-
```

Comment:
Why does this test have to go?

Reply:
> Why does this test have to go?

@cgcgcg, see https://github.com/trilinos/Trilinos/pull/8115/commits/b14e80f44384fe180d738730998b3da2a1922b0a.

It looks to be a complete copy-and-paste duplicate.  Take a look at this file in the current 'develop' branch showing the duplicate tests at lines:

*  https://github.com/trilinos/Trilinos/blob/develop/packages/muelu/test/toggletransfer/CMakeLists.txt#L88
* https://github.com/trilinos/Trilinos/blob/develop/packages/muelu/test/toggletransfer/CMakeLists.txt#L96

showing:

```
  TRIBITS_ADD_TEST(
    TogglePFactory_driver
    NAME "Driver_TogglePFactory_semi_sa_line_easy_Tpetra"
    ARGS "--matrixType=Laplace3D --nx=40 --ny=40 --nz=60 --mz=1 --xmlFile=parameters_semi_sa_easy.xml --linAlgebra=Tpetra"
    NUM_MPI_PROCS 4
    COMM serial mpi
    )

  TRIBITS_ADD_TEST(
    TogglePFactory_driver
    NAME "Driver_TogglePFactory_semi_sa_line_easy_Tpetra"
    ARGS "--matrixType=Laplace3D --nx=40 --ny=40 --nz=60 --mz=1 --xmlFile=parameters_semi_sa_easy.xml --linAlgebra=Tpetra"
    NUM_MPI_PROCS 4
    COMM serial mpi
    )
```

That looks like an exact char-for-char copy.

Does it not look that way to you?

Reply:
Also, note that it makes no sense to specify `COMM serial` when the test requires more than one MPI proc.  That test will just get disabled in a serial build and print a note about it not getting added.  You can see that at [here](https://testing.sandia.gov/cdash/build/8040544/configure) showing:

```
-- MueLu_Driver_TogglePFactory_semi_sa_line_easy_Tpetra: NOT added test because NUM_MPI_PROCS='4' > MPI_EXEC_MAX_NUMPROCS='1'!
```

Make sense?

Reply:
Oh boy..
From the git log, it looks like one of them was supposed to do Epetra and the other Tpetra. I'm cool with your change, but @trilinos/muelu will have a look at these tests. And the `NUM_MPI_PROCS 4` `COMM serial mpi` pattern seems to be all over that file.

Reply:
> And the `NUM_MPI_PROCS 4` `COMM serial mpi` pattern seems to be all over that file.

@cgcgcg, this is harmless but a bit confusing. (It makes no sense to define a 4 MPI proc test to run without MPI which is what one is saying when they explicitly pass in `COMM serial ...`.)

Note that if you don't specify a `COMM` argument, by default the test will be defined independent of the value of `TPL_ENABLE_MPI`  See:

* https://tribits.org/doc/TribitsDevelopersGuide.html#formal-arguments-tribits-add-test

Reply:
Hm, I'm confused now. In this serial build https://testing.sandia.gov/cdash/viewTest.php?onlypassed&buildid=8041837 the tests are running. So specifying `COMM serial mpi` and `MPI_NUM_PROCS 4` seems to be running if `(mpi and 4 procs) or serial`.

Reply:
@cgcgcg,

This is confusing either way you slice it I guess.  The old behavior (before this PR) was to ignore the `NUM_MPI_PROCS` argument for non-MPI builds and essentially make it 1.  But that is not good to do for the reason described below.

What most packages do (or at least the one's I have contributed to) is to define one test that explicitly uses a single process (that will always run with MPI or non-MPI builds) and define an extra test that uses more than one MPI process for MPI builds like this:

```
 TRIBITS_ADD_TEST(
    TogglePFactory_driver
    NAME "Driver_TogglePFactory_semi_sa_line_easy_Tpetra"
    ARGS "--matrixType=Laplace3D --nx=40 --ny=40 --nz=60 --mz=1 --xmlFile=parameters_semi_sa_easy.xml --linAlgebra=Tpetra"
    NUM_MPI_PROCS 1
    )

 TRIBITS_ADD_TEST(
    TogglePFactory_driver
    NAME "Driver_TogglePFactory_semi_sa_line_easy_Tpetra"
    ARGS "--matrixType=Laplace3D --nx=40 --ny=40 --nz=60 --mz=1 --xmlFile=parameters_semi_sa_easy.xml --linAlgebra=Tpetra"
    NUM_MPI_PROCS 4
    COMM mpi
    )
```

What that does is to define a test `MueLu_Driver_TogglePFactory_semi_sa_line_easy_Tpetra_MPI_1` (called  `MueLu_Driver_TogglePFactory_semi_sa_line_easy_Tpetra` in a non-MPI build) that will run in both MPI and non-MPI builds.  And in MPI builds, the test `MueLu_Driver_TogglePFactory_semi_sa_line_easy_Tpetra_MPI_4` will also get defined and run.

You don't typically want to set up your test suite to so there are tests that only run in a non-MPI build.  Instead, you want to set up your test suite so that the tests that run in a non-MPI build are a strict subset of the tests that run in an MPI build.  That way, you can just do an MPI build and have confidence that all of the tests will still pass for a non-MPI build.

Does that make sense?

I don't know if that strategy is written down anywhere but it should be if it is not.

Reply:
So that means that MueLu will have to redefine all of its tests? I wish I had reviewed this PR more carefully..

Reply:
> So that means that MueLu will have to redefine all of its tests? I wish I had reviewed this PR more carefully..

@cgcgcg, only if you want.  For MPI builds, zero testing has been lost.

But does that mean that MueLu does not have sufficient testing with a single MPI process in an MPI build?  That is the real issue here I think.

Reply:
No, I think that MueLu just lost all its tests for serial builds, since it relied on the previous behavior. MPI tests should be fine.

Reply:
Would it be possible to add a flag to `TRIBITS_ADD_TEST` that allows us to get the previous behavior back, instead of duplicating all of our tests for serial?

Reply:
> Would it be possible to add a flag to TRIBITS_ADD_TEST that allows us to get the previous behavior back, instead of duplicating all of our tests for serial?

I think a better approach that would also provide for single MPI rank tests in an MPI build would be to define and use a function `MUELU_ADD_TEST( ... NUM_MPI_PROCS <numMpiProcs> ... )` that defines two tests for an MPI build (one with 1 MPI rank and one with  `<numMpiProcs>` ranks if  `<numMpiProcs> > 1`) and a single test for non-MPI builds.   That could be applied with a sed script to update the MueLu CMakeLists.txt files.

Reply:
@cgcgcg, I can help you write and apply that function if you are interested.

Reply:
Yep, that's also an option. Let's take this offline, I'll contact you.